src/server/graphql/mutations/__tests__/removeTeamMember.test.js

```diff
@@ -0,0 +1,88 @@
+import MockDate from 'mockdate';
+import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
+import {__now} from 'server/__tests__/setup/mockTimes';
+import removeTeamMember from 'server/graphql/mutations/removeTeamMember';
+import MockDB from 'server/__tests__/setup/MockDB';
+import MockPubSub from 'server/__mocks__/MockPubSub';
+import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
+import getRethink from 'server/database/rethinkDriver';
+import {auth0ManagementClient} from 'server/utils/auth0Helpers';
+import DynamicSerializer from 'dynamic-serializer';
+import * as getPubSub from 'server/utils/getPubSub';
+import expectAsyncToThrow from 'server/__tests__/utils/expectAsyncToThrow';
+
+MockDate.set(__now);
+console.error = jest.fn();
+
+describe('removeTeamMember', () => {
+  test('promotes another member if the person removed was the lead', async () => {
+
+  });
+
+  test('removes the teamMember, reassigns active projects to the lead', async () => {
+    // SETUP
+    const r = getRethink();
+    const mockPubSub = new MockPubSub();
+    getPubSub.default = () => mockPubSub;
+    const mockDB = new MockDB();
+    const dynamicSerializer = new DynamicSerializer();
+    await mockDB
+      .init()
+      .newProject();
+    auth0ManagementClient.__initMock(mockDB.db);
+    auth0ManagementClient.users.updateAppMetadata.mockReset();
+    const userToBoot = mockDB.db.user[7];
+    const authToken = mockAuthToken(mockDB.db.user[0]);
+    const teamId = mockDB.db.team[0].id;
+    // TEST
+    const teamMemberId = mockDB.db.teamMember[7].id;
+    await removeTeamMember.resolve(
+      undefined,
+      {teamMemberId},
+      {authToken}
+    );
+
+    const db = await fetchAndSerialize({
+      teamMember: r.table('TeamMember').getAll(teamId, {index: 'teamId'}).orderBy('checkInOrder'),
+      project: r.table('Project').getAll(teamId, {index: 'teamId'}),
+      user: r.table('User').get(userToBoot.id)
+    }, dynamicSerializer);
+    expect(db).toMatchSnapshot();
+    expect(auth0ManagementClient.users.updateAppMetadata.mock.calls.length)
+      .toBe(1);
+    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot();
```

Comment:
I wonder if this snapshot boilerplate is saving us that much effort as a pattern. Calling `expect(db).toMatchSnapshot();` is useful when our test has effected a lot of state, but the tradeoff is that the code is more implicit.

This feels like a case in which we can just do something like `theProjectOwners.forEach((owner) => expect(owner).toEqual(theTeamLead));`

Reply:
for single fields, absolutely! but take the pub sub for example:
```
 +              "isKickout": true,
 +              "orgId": 26,
 +              "startAt": 2017-01-01T08:00:00.000Z,
 +              "teamId": 16,
 +              "teamName": "Team America",
 +              "type": "KICKED_OUT",
 +              "userIds": Array [
 +                24,
 +              ],
```
await r({user: r.table('User').get(userId), team: r.table('Team).get(teamId))......})`
expect(notification.teamId).toEqual(team.id)
expect(notification.teamName).toEqual(team.preferredName)
expect(notification.type).toEqual('KICKED_OUT')
expect(userIds).toEqual([user.id])
```
_even_ if you wrote these 5 tests, you're still not protecting yourself from a break. someone could bork the `startAt` date, or they could add an extra `password` field.
At the end of the day, you end up writing 20-30 expectations, 1 for every field, and then business logic changes & you have to completely rewrite them. so much nicer to say "hey, here's a picture. if that picture changes, tell me".

Reply:
I think we agree. The tradeoff definitely operates around the question of "am I writing way too many expectations to verify the expected behavior?"

That said, I think the other end of the tension is "Am I testing unrelated or out-of-scope side effects in the database?" An argument can be made that, at the unit-test layer, the pub-sub behavior is not what we necessarily need to be testing; rather just that the user was really removed from the team and that their projects were reassigned.

Another problem I see with snapshot testing, is that, as the person reading said tests, it can be difficult to read the actual expected behavior. Reading a database dump isn't quite the same as reading `expect(theReturnValue).toEqual(whatIsExpected)`. We can probably ease this tradeoff by adding code comments stating the expected behavior.

Since I'm new to snapshot testing and this code base, though, and cause the test looks correct I'm not gonna request that you change it. Just shooting around some thoughts

Reply:
> An argument can be made that, at the unit-test layer, the pub-sub behavior is not what we necessarily need to be testing

If I call `removeTeamMember`, i expect it to tell that team member that they got kicked off the team. if it doesn't send that message, it's broken. Is there a better place to test that functionality?

> Another problem I see with snapshot testing, is that, as the person reading said tests, it can be difficult to read the actual expected behavior

Absolutely agree! so a snapshot test is exactly same as `expect(theReturnValue).toEqual(whatIsExpected)` but `whatIsExpected` is an object stored in another file. Personally, i'd rather make 1 call to compare an object instead of 30 calls on each primitive, but that's just me :-)

Reply:
> If I call removeTeamMember, i expect it to tell that team member that they got kicked off the team. if it doesn't send that message, it's broken. Is there a better place to test that functionality?

This is a super good point, and something which I struggle with when coming up with test plans. You're obviously correct that when you call that mutation, you want to publish a message on a channel. I'd say that what we're getting at here is the distinction between a unit test and an integration test. If it were a unit test, I'd check that it updated the DB correctly. For an integration level test, I'd want to make sure that _when_ I update certain documents/fields in the DB, messages are published appropriately.

> Personally, i'd rather make 1 call to compare an object instead of 30 calls on each primitive, but that's just me :-)

Hey, that's not just you - I agree that I'd never write 30 assertions in a test! I'm just pointing out that we should generally avoid writing tests that require us to test 30 side effects / impact on the whole database.

Reply:
great points! i suppose to be super specific i'd call this an "endpoint test". if it were a unit test, i'd be testing the helper function `removeAllTeamMembers`, which is called by both "removeOrgMember" and "removeTeamMember". By testing `removeAllTeamMembers` we'd be focusing on _how_ it works (ie does it use `removeAllTeamMembers()`?) All i care about is the end result: a new record in the DB & a new message in the pubsub.

Reply:
Hey, likewise, really digging your perspective here. I'd actually take your following point even further:

> By testing removeAllTeamMembers we'd be focusing on how it works (ie does it use removeAllTeamMembers()?) All i care about is the end result: a new record in the DB & a new message in the pubsub.

Ok, yeah! Agree. I'd say we should always strive to test the _what_ (behavior) over the _how_ (implementation). Even though unit tests are more granular, it's always good to be thinking about the behavior of the current level you're testing at. If the behavior of a small unit is to call another small unit, then that may count as the _what_ at that level, even if it's a _how_ at a higher level of abstraction.

:)